src/backend/postmaster/loginmonitor.c

```diff
@@ -0,0 +1,925 @@
+
+	/*
+	 * Force zero_damaged_pages OFF in the login monitor process, even if it is set`
+	 * in postgresql.conf. We don't really want such a dangerous option being applied
+	 * non-interactively.
+	 */
+	SetConfigOption("zero_damaged_pages", "false", PGC_SUSET, PGC_S_OVERRIDE);
+
+	/*
+	 * Force settable timeouts off to avoid letting these settings prevent
+	 * regular maintenance from being executed.
+	 */
+	SetConfigOption("statement_timeout", "0", PGC_SUSET, PGC_S_OVERRIDE);
+	SetConfigOption("lock_timeout", "0", PGC_SUSET, PGC_S_OVERRIDE);
+	SetConfigOption("idle_in_transaction_session_timeout", "0",
+			PGC_SUSET, PGC_S_OVERRIDE);
+
+	/*
+	 * Force default_transaction_isolation to READ COMMITTED. We don't want
+	 * to pay the overhead of serializable mode, nor add any risk of causing
+	 * deadlocks or delaying other transactions.
+	 */
+	SetConfigOption("default_transaction_isolation", "read committed",
+			PGC_SUSET, PGC_S_OVERRIDE);
+
+	LoginMonitorShmem->lm_pid = MyProcPid;
+	LoginMonitorShmem->lm_latch = MyLatch;
+
+	/*
+	 * Main loop until shutdown request
+	 */
+	while (!ShutdownRequestPending) {
+		/*
+		 * Wait until naptime expires or we get some type of signal (all the
+		 * signal handlers will wake us by calling SetLatch).
+		 */
+		(void) WaitLatch(MyLatch,
+				 WL_LATCH_SET | WL_TIMEOUT | WL_EXIT_ON_PM_DEATH,
+				 5 * 1000L,
+				 WAIT_EVENT_LOGIN_MONITOR_LAUNCHER_MAIN);
+
+		/* Clear any already-pending wakeups */
+		ResetLatch(MyLatch);
+
+		HandleLoginMonitorInterrupts();
+
+		/*
+		 * As only one worker can run at any time, before start a worker,
+		 * we should check whether there is already one worker running.
+		 */
+		if (LoginMonitorShmem->login_failed_requested)
+		{
+			SendPostmasterSignal(PMSIGNAL_START_LOGIN_MONITOR_WORKER);
+			elog(DEBUG1, "Login monitor launcher is processing uesr %s and has sent starting"
+	     					"worker signal to postmaster.", LoginMonitorShmem->curr_user_name);
+			WaitLatch(LoginMonitorShmem->lm_latch,
+				  WL_LATCH_SET | WL_POSTMASTER_DEATH, 0,
+				  WAIT_EVENT_LOGINMONITOR_FINISH);
+			/* Clear any already-pending wakeups */
+			ResetLatch(LoginMonitorShmem->lm_latch);
+		}
+	}
+
+	LoginMonitorShutdown();
+}
+
+/********************************************************************
+ *				LOGIN MONITOR WORKER CODE
+ ********************************************************************/
+#ifdef EXEC_BACKEND
+/*
+ * forkexec routines for the login monitor worker.
+ *
+ * Format up the arglist, then fork and exec.
+ */
+static pid_t
+lmworker_forkexec(void)
+{
+	char	*lm[10];
+	int			ac = 0;
+
+	lm[ac++] = "postgres";
+	lm[ac++] = "--forklmworker";
+	lm[ac++] = NULL;		/* filled in by postmaster_forkexec */
+	lm[ac] = NULL;
+
+	Assert(ac < lengthof(lm));
+
+	return postmaster_forkexec(ac, lm);
+}
+
+/*
+ * We need this set from the outside, before InitProcess is called
+ */
+void
+LoginMonitorWorkerIAm(void)
+{
+	am_login_monitor_worker = true;
+}
+#endif
```

Comment:
What's the usage of this function?

Reply:
This code is used to compatibility of WIN platform, which now GP DB and Cloudberry DB is not supported.  I can't confirm whether to remove that.